src/merkle/store/mod.rs

```diff
@@ -244,6 +244,58 @@ impl MerkleStore {
         })
     }
 
+    /// Reconstructs a path from the root until a leaf or empty node and returns its depth.
+    ///
+    /// The empty node is always calculated for maximum depth of `64`.
+    ///
+    /// # Errors
+    /// Returns an error if the provided root is not found, or if the path opens to a tree with
+    /// depth greater than `64`.
+    pub fn get_leaf_depth(&self, root: Word, index: u64) -> Result<u8, MerkleError> {
+        // the path can go up to 64 bits; hence, this is the maximum depth
+        const MAX_DEPTH: u8 = u64::BITS as u8;
+        let empty = EmptySubtreeRoots::empty_hashes(MAX_DEPTH);
+
+        // check if the root exists, providing the proper error report if it doesn't
+        let mut hash: RpoDigest = root.into();
+        self.nodes
+            .get(&hash)
+            .ok_or(MerkleError::RootNotInStore(hash.into()))?;
+
+        // we traverse from root to leaf, so the path is reversed
+        let mut path = index.reverse_bits();
+
+        // iterate every depth and reconstruct the path from root to leaf
+        for depth in 0..MAX_DEPTH {
+            // we short-circuit if an empty node has been found
+            if hash == empty[depth as usize] {
+                return Ok(depth);
```

Comment:
is this correct? that is the depth of the first inner node that contains only empty leaf nodes, but that is not the same as the depth of the empty leaf.

For example, if this method is called with the empty root, this would match on first iteration and the reported depth would be `0`, but the empty leaf should be at depth 64, no?

Reply:
We always check the current depth here as we set the hash of the next depth by the end of the for loop.

Example:

starts with `hash: root, depth: 0`
next iter `hash: root child (left or right), depth: 1`
next iter `hash next child (left or right), depth: 2`
etc

Reply:
I don't mean the depth that is being checked, I mean the depth that is being returned. It seems it returns the depth of the first inner node that contains only empty leaves, but the name of the method says it returns the depth of the leaf itself

Reply:
Let's assume we have the following tree

```
   a
  / \
 c   b
/ \ / \
d e f g
```

and we have a path `01` that is `a, c, e`

The first iteration will be `depth: 0, hash: a, left: c, right: b`. We check if `a == empty`, and return `0` in that case.

Then we move hash to `c` and increment the depth by `1`. Since that check is the first thing we do, if `hash == empty`, or if `c == empty`, return `1`. Then we inc again and move to `e`

Then, if `e == empty`, we return `2`.

Reply:
> We check if a == empty, and return 0 in that case.

that is what I mean, `a` is the `root`, not a leaf. But it seems that is the behavior you want.

Reply:
Hmm it's not really a vacant node. The definition of a leaf we are using is "a node without children", with the exception of an empty node that will be treated as if it doesn't have children.

IMO the name `get_leaf_depth` is comprehensive enough, as we will have an empty node only if a leaf/node was never inserted there, but I don't oppose using different names. But in principle, you are right, this is a sort of "loose" leaf definition as we do have children for an empty root (i.e. the next depth empty sub-tree roots).

wdyt @bobbinth ?